daemon/src/taker_cfd.rs

```diff
@@ -86,20 +87,25 @@ impl Actor {
                 short
             });
 
-            (
-                maybe_contract_symbol.unwrap_or(ContractSymbol::BtcUsd),
-                maker_offers,
-            )
+            maker_offers
         });
 
         self.current_maker_offers = takers_perspective_of_maker_offers.clone();
         tracing::trace!("new maker offers {:?}", takers_perspective_of_maker_offers);
 
+        // If there's no offer, it might as well be hard-coded to BTC
+        let symbol = self
+            .current_maker_offers
+            .clone()
+            .and_then(|offer| offer.long.map(|offer| offer.contract_symbol))
+            .unwrap_or(ContractSymbol::BtcUsd);
```

Comment:
🤔 Surely this is wrong. We are getting an update from the maker about a particular type of offer here. We can't just assume that it's `BTCUSD` just because we don't know the contract symbol! AFAIK we need to use this information properly in order to disable certain offers.

I think we need to rework this and not just build on top of it. The `MakerOffers` type just doesn't work for what we are trying to do. I think something like this could work better:

```rust
struct MakerOfferInfo {
    pub contract_symbol: ContractSymbol,
    long: Option<OfferParams>,
    short: Option<OfferParams>,
    fee_params: Option<FeeParams>,
}

struct OfferParams {
// all the fields in `Offer` minus `contract_symbol`,
// `position_maker`, `tx_fee_rate`, and `funding_rate`
}

struct FeeParams {
    tx_fee_rate: TxFeeRate,
    funding_rate_long: FundingRate,
    funding_rate_short: FundingRate,
}

impl MakerOfferInfo {
    // based on `self.long` and `self.fee_params`
    fn long_offer(&self) -> Option<Offer>;

    // based on `self.short` and `self.fee_params`
    fn short_offer(&self) -> Option<Offer>;

    fn fee_params(&self) -> Option<FeeParams>;
}
```

This should also give us maximum flexibility to allow rollovers whilst disabling new orders if we want to. And keeping all fields orthogonal helps us avoid inconsistencies and makes this type clearer.

Reply:
We unfortunately can't easily remove `contract_symbol` from `Offer` because that's being sent over the wire( backwards compatibility). 
I could try to get rid of `MakerOffers` and split out `FundingRate` but that's a bigger change. I'll give this a try

Reply:
No, doesn't work easily because we send `MakerOffers` over the wire ... :/

Reply:
Actually, the protocol does not work with multiple symbols because it assumes `LatestMakerOffers` which has an `Option<MakerOffers>`. If this inside is `None`, it would set the offers inside the taker to `None`. Hence, `LatestMakerOffers` would need a field `contract_symbol`. 

This should be the scope of https://github.com/itchysats/itchysats/issues/2459 and changing this here would kind of explode in scope. 

What do you think?

Reply:
We discussed this offline and the conclusion was:

- The current state of this PR is not ideal, but we can basically merge it given that we are gonna tackle the concerns I raised with #2459.
- @bonomat is gonna try to split the `MakerOffers` into two types: one for the offer protocol (single contract symbol) and one for the rest of the application (multiple contract symbols).

Reply:
I'm afraid removing the `contract_symbol` from `Offer` is not easy. We rely on this information when placing an order: the taker places the order only using the `order id` with which we load the `offer` and create a `Cfd`. For creating the `Cfd` we need to know the `contract_symbol` which is not available there....

Reply:
Maybe there was some miscommunication. I don't think we should remove the `ContractSymbol` from the `Offer`! My proposal was to introduce a _new_ `OfferParams` type that is `ContractSymbol`-agnostic. One can then compute an `Offer` from the proposed `MakerOfferInfo` only if all the corresponding parameters are available, by calling either `long_offer` or `short_offer`.

Does that make sense?

Reply:
> Does that make sense?

Not yet. What's the advantage of this if we still need to keep the old struct `Offer` around?

Reply:
I'm working under the assumption that `Offer` is a type that we can't easily modify because of DB compatibility. Furthermore, and offer is always about a particular contract symbol, so there's nothing wrong with the type.

The problem that I see is that the thing that we send around for the `/itchysats/offer` protocol, `Option<MakerOffers>` doesn't work well with multiple contract symbols:

- If we send `None`, the receiver doesn't know what we are talking about.
- If we send `MakerOffers` and the `long` and `short` offers are set `None` (in order to disable the creation of new orders), the receiver doesn't know how to use the `tx_fee_rate`, `funding_rate_long` and `funding_rate_short` fields for rollovers, because it's missing the contract symbol context.
- The `long` and `short` offer fields aren't guaranteed to be talking about the same contract symbol, which is confusing and could lead to errors. This highlights the importance of extracting the contract symbol into a common place.

That's why I propose to do something like what I said in https://github.com/itchysats/itchysats/pull/2615#discussion_r936668903. The advantages I see are:

- The taker will always know the contract symbol of the offer information they're receiving from the maker.
- The contract symbol at the top level of the struct is the only source of truth. Because the proposed `OfferParams` types don't talk about the contract symbol, they can never disagree with the top level. Consumers cannot access these inner fields and instead we provide `long_offer` and `short_offer` methods that reconstruct the `Option<Offer>` type based on what's available. The `Offer` type therefore doesn't have to change.
- If the proposed field `fee_params` is `None`, we can disable new orders and rollovers. 

We might want to move something like the `creation_timestamp` to the top level as well, if we want to be able to consider all the offer information stale.

Reply:
> The problem that I see is that the thing that we send around for the /itchysats/offer protocol, Option<MakerOffers> doesn't work well with multiple contract symbols:

Are you _only_ talking about the types send over the network? 
I was talking about the _internal_ structure we are using from the http api of the maker throughout `projection` and other actors. 

I think the network layer can easily be cleaned up as you propose. That's what I prepared with [`a7d2335` (#2615)](https://github.com/itchysats/itchysats/pull/2615/commits/a7d2335e9c7de27151901c125d7e33b9b11114af)  and would tackle in #2459 . The internal part is a bit more work though but once the network layer is set in stone we can shuffle things around internally easily.

Reply:
Yeah, I was certainly focusing on these types in the context of the network layer.